Add documentation (#8)
hm-aemy/circt-pass-plugin-example: README.md

> This repository aims to provide a project structure and CMake configuration that enables users to write CIRCT passes as standalone plugins outside of CIRCT's regular compilation tree.

Reviewer: Let's just start a bit more purpose full, something like

```suggestion
Here is an example to build a [CIRCT](github.com/llvm/circt) standalone pass plugin.
This out-of-tree approach makes it possible ...
```

CIRCT is part of LLVM's MLIR project. While this repository focuses on the CIRCT project, it can also be used to develop MLIR plugins.

# Table of Contents

> # Table of Contents

Reviewer: ```suggestion
```

- [CIRCT Plugin](#circt-plugin)
- [Components of a plugin](#components-of-a-plugin)
    - [Specification of the pass](#specification-of-the-pass)
    - [Pass logic](#pass-logic)
    - [Pass registration](#pass-registration)
- [Requirements](#requirements)
- [Implementation](#implementation)
  - [Updating the naming](#updating-the-naming)
  - [Updating the TableGen description](#updating-the-tablegen-description)
  - [Implementing the pass](#implementing-the-pass)
  - [Pass registration](#pass-registration-1)
- [Compilation](#compilation)
- [Execution](#execution)
- [Next steps](#next-steps)

#  CIRCT Plugin

> Usually, CIRCT passes reside within its [sources](https://github.com/llvm/circt), usually as part of a dialect, and are compiled as a part of CIRCT.

Reviewer: drop the "usually"s

> This out-of-tree approach enables to compile a pass separately as a plugin and load it into CIRCT later; for example with ```circt-opt```.

Reviewer: ```suggestion
This out-of-tree approach enables to build a pass separately as a plugin and use it with CIRCT later; for example with ```circt-opt```.
```

This repository contains the project structure and CMake config for such a plugin along with a guide on
- what components a CIRCT plugin consists of and have to be implemented / adapted by the user,
- how a plugin is compiled using CIRCT/MLIR/LLVM as an external source and

> - how the plugin can be run using ```circt-opt```/```mlir-opt``` to operate on ```.mlir``` files.

Reviewer: ```suggestion
- how the plugin can be run using `circt-opt`/`mlir-opt` to operate on `.mlir` files.
```

I think here, and also at other places, it's better to use only single '`'

> # Components of a plugin

Reviewer: My feeling is that is nicer to only use H1 one for the repo name (not in this readme present) and then use H2 for the main headers.

> Below is a brief overview of the components that make up a CIRCT/MLIR plugin. A more detailed description on how to work with these is provided under [Implementation](#Implementation)

Reviewer: Since we already have the note a the top that this should work analogous for MLIR, I would not repeat it in the text.

Reviewer: Also the dot is missing at the end.

## Specification of the pass
MLIR uses the [TableGen](https://mlir.llvm.org/docs/PassManagement/#tablegen-specification) (```.td```) format for "Declarative Specification" of passes, operations etc., based on which the relevant C++ boilerplate is generated.
TableGen is used for many things like defining the legal input- and output dialects of a pass or defining which operation types the pass runs on.

> This guide will only cover the bare minimum required to generated the boilerplate necessary for a simple plugin.

Reviewer: ```suggestion
This guide will only cover the bare minimum required to generate the code necessary for a simple plugin.
```

## Pass logic

> Based on the TableGen specification, MLIR generates a C++ base class with the necessary boilerplate code. The actual pass logic resides in a struct that inherits from this base class, invoking relevant MLIR/CIRCT functionlity where needed.

Reviewer: This first sentence sounds a bit redundant to the previous section. Maybe make it a bit shorter and focus on the second part.

## Pass registration
To use the plugin, we have to "plug it into" the MLIR infrastructure like MLIR's pass management. This is done via pass registration.
Registering a pass as a plugin requires exposing a standard interface discoverable by MLIR, which provides metadata about the plugin and acts as an entry point from which the pass registration can be performed.

# Requirements
To write a plugin for CIRCT/MLIR, you need CIRCT/MLIR itself, which in turn requires LLVM.

> > **Note:** This guide will use the included submodule but if you already have an installation of CIRCT/MLIR/LLVM you can use that by changing the ```CIRCT_DIR```, ```MLIR_DIR``` and ```LLVM_DIR``` variables in the ```CmakeLists.txt``` file.

Reviewer: also the submodules are not required
```suggestion
> **Note:** This guide will use the included submodule but if you already have an installation of CIRCT/MLIR/LLVM you can use that by changing the ```CIRCT_DIR```, ```MLIR_DIR``` and ```LLVM_DIR``` variables in the ```CMakeLists.txt``` file.
```
also the variables have not to be changed, but rather set by cmake invocation, as noted below.

Reviewer: add a note that it will work with the git version from the submodules and might work with newer upstream version, but this is then not guaranteed.

    # configure cmake
    cmake -G Ninja ../llvm \
        -DLLVM_ENABLE_PROJECTS="mlir" \
        -DLLVM_TARGETS_TO_BUILD="host" \
        -DLLVM_ENABLE_ASSERTIONS=ON \
        -DCMAKE_BUILD_TYPE=DEBUG \
        -DLLVM_USE_SPLIT_DWARF=ON \
        -DLLVM_ENABLE_LLD=ON
    # compile
    ninja
    cd ../..
    ```

- **Build CIRCT:**
    > **Note:** If you want to use this repository to develop an MLIR-plugin without CIRCT, ignore this step.
    ``` sh
    # cwd: circt-pass-plugin-example/circt/
    mkdir build && cd build
    cmake -G Ninja .. \
        -DMLIR_DIR=$PWD/../llvm/build/lib/cmake/mlir \
        -DLLVM_DIR=$PWD/../llvm/build/lib/cmake/llvm \
        -DLLVM_ENABLE_ASSERTIONS=ON \
        -DCMAKE_BUILD_TYPE=DEBUG \
        -DLLVM_USE_SPLIT_DWARF=ON \
        -DLLVM_ENABLE_LLD=ON \
        -DCIRCT_SLANG_FRONTEND_ENABLED=ON
    ninja
    cd ../..
    ```

With this done, you have the necessary dependencies installed and can now go about writing your pass plugin.

# Implementation
To implement a plugin, the previously described components have to be realized. This section covers the necessary changes and their locations within this project.

## Updating the naming

> To set the naming of your pass, replace the placeholder-names throughout this repository. If you were to choose ```MyName``` for example, that would mean changing all occurrences of ```CustomOperation``` to ```MyName```, ```CustomOperationPass``` to ```MyNamePass``` and so on within the following files, also changing the file- and directory-names:

Reviewer: I think it might be good to state that this repo for now contains only the structure and not really a functional example which is doing something at the top. Maybe explain that it can be build as is, but the plugin will not do anything, except being present and able to be loaded.

- ```include/CustomOperation/CustomOperation.td```
- ```include/CustomOperation/CustomOperation.h```
- ```include/CustomOperation/CmakeLists.txt```
- ```include/CmakeLists.txt```
- ```src/CustomOperation.cpp```
- ```CmakeLists.txt```

The ```CustomOperation``` naming scheme will still be used throughout this ```README```.

## Updating the TableGen description
Within ```include/CustomOperation/CustomOperation.td```, there are two strings you use to provide a description of what your pass does.
In the field ```summary``` you can give a brief description, while you use the ```description``` field to elaborate further.

> This is also the place to extend the specification of the pass, as described in [Specification of the pass](#Specification%0aof%0athe%0apass).

Reviewer: left over from your local markdown rendering?

## Implementing the pass
The actual logic of your pass resides in a struct that inherits from the base class generated from the TableGen description.
Within ```src/CustomOperation.cpp```, you find the following lines:
``` cpp
struct CustomOperation : public impl::CustomOperationPassBase<CustomOperation> {
  void runOnOperation() override {
      // TODO: implement pass-logic here
  }
};
```
Here, ```runOnOperation()``` serves as the entry point through which MLIR delegates control when running your pass.

> For more information on implementing a pass, refer to the official [CIRCT](https://circt.llvm.org/docs/Passes/) or [MLIR](https://mlir.llvm.org/docs/PassManagement/) documentation or look at the examples mentioned under [Next steps](#Next%0asteps).

Reviewer: link again
